File: go/v2/transactions.go
package v2

import (
	"bytes"
	"context"
	"fmt"
	"strconv"
	"time"

	tesote "github.com/tesote/sdk/go"
)

// TransactionsService groups v2 transaction endpoints.
type TransactionsService struct {
	client *tesote.Client
}

// TransactionsFilter holds the shared filter parameters for v2 transaction queries.
type TransactionsFilter struct {
	StartDate             string
	EndDate               string
	Scope                 string
	Page                  int
	PerPage               int
	TransactionsAfterID   string
	TransactionsBeforeID  string
	TransactionDateAfter  string
	TransactionDateBefore string
	CreatedAfter          string
	UpdatedAfter          string
	AmountMin             string
	AmountMax             string
	Amount                string
	Status                string
	CategoryID            string
	CounterpartyID        string
	Q                     string
	Type                  string
	ReferenceCode         string
}

func (f TransactionsFilter) query() map[string]string {
	q := map[string]string{}
	set := func(k, v string) {
		if v != "" {
			q[k] = v
		}
	}
	set("start_date", f.StartDate)
	set("end_date", f.EndDate)
	set("scope", f.Scope)
	if f.Page > 0 {
		q["page"] = strconv.Itoa(f.Page)
	}
	if f.PerPage > 0 {
		q["per_page"] = strconv.Itoa(f.PerPage)
	}
	set("transactions_after_id", f.TransactionsAfterID)
	set("transactions_before_id", f.TransactionsBeforeID)
	set("transaction_date_after", f.TransactionDateAfter)
	set("transaction_date_before", f.TransactionDateBefore)
	set("created_after", f.CreatedAfter)
	set("updated_after", f.UpdatedAfter)
	set("amount_min", f.AmountMin)
	set("amount_max", f.AmountMax)
	set("amount", f.Amount)
	set("status", f.Status)
	set("category_id", f.CategoryID)
	set("counterparty_id", f.CounterpartyID)
	set("q", f.Q)
	set("type", f.Type)
	set("reference_code", f.ReferenceCode)
	return q
}

// ListForAccount lists transactions for an account. GET /v2/accounts/{id}/transactions.
func (s *TransactionsService) ListForAccount(ctx context.Context, accountID string, filter TransactionsFilter) (*tesote.TransactionListResponse, error) {
	out := &tesote.TransactionListResponse{}
	_, err := s.client.Do(ctx, "GET", pathPrefix+"/accounts/"+accountID+"/transactions", tesote.RequestOptions{
		Query:    filter.query(),
		Out:      out,
		CacheTTL: time.Minute,
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

// Get fetches a single transaction. GET /v2/transactions/{id}.
func (s *TransactionsService) Get(ctx context.Context, id string) (*tesote.Transaction, error) {
	out := &tesote.Transaction{}
	_, err := s.client.Do(ctx, "GET", pathPrefix+"/transactions/"+id, tesote.RequestOptions{
		Out:      out,
		CacheTTL: 5 * time.Minute,
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

// ExportFormat is the wire enum for /v2/accounts/{id}/transactions/export `format`.
type ExportFormat string

// Supported export formats.
const (
	ExportFormatCSV  ExportFormat = "csv"
	ExportFormatJSON ExportFormat = "json"
)

// ExportOptions tunes GET /v2/accounts/{id}/transactions/export.
type ExportOptions struct {
	Filter TransactionsFilter
	Format ExportFormat
}

// ExportResult is the raw export payload (CSV or pretty JSON bytes).
type ExportResult struct {
	Body        []byte
	ContentType string
	Filename    string
}

// Export downloads transactions as CSV or JSON.
// GET /v2/accounts/{id}/transactions/export. Returns the raw payload bytes.
func (s *TransactionsService) Export(ctx context.Context, accountID string, opts ExportOptions) (*ExportResult, error) {
	q := opts.Filter.query()
	if opts.Format != "" {
		q["format"] = string(opts.Format)
	}
	resp, err := s.client.Do(ctx, "GET", pathPrefix+"/accounts/"+accountID+"/transactions/export", tesote.RequestOptions{
		Query: q,
	})
	if err != nil {
		return nil, err
	}
	res := &ExportResult{
		Body:        append([]byte(nil), resp.Body...),
		ContentType: resp.Header.Get("Content-Type"),
	}
	if cd := resp.Header.Get("Content-Disposition"); cd != "" {
		// why: extract filename from Content-Disposition without bringing in mime/multipart.
		res.Filename = parseFilename(cd)
	}
	return res, nil
}

// parseFilename pulls a filename from a Content-Disposition header (best-effort).
func parseFilename(cd string) string {
	const marker = "filename="
	idx := bytes.Index([]byte(cd), []byte(marker))
	if idx < 0 {
		return ""
	}
	rest := cd[idx+len(marker):]
	if len(rest) > 0 && rest[0] == '"' {
		end := bytes.IndexByte([]byte(rest[1:]), '"')
		if end < 0 {
			return rest[1:]
		}
		return rest[1 : 1+end]
	}
	end := bytes.IndexAny([]byte(rest), "; ")
	if end < 0 {
		return rest
	}
	return rest[:end]
}

// SyncOptions tunes the body of POST /v2/accounts/{id}/transactions/sync (and legacy).
type SyncOptions struct {
	Count          int
	Cursor         string
	Options        SyncSubOptions
	IdempotencyKey string
}

// SyncSubOptions is the nested `options` block on a sync request.
type SyncSubOptions struct {
	IncludeRunningBalance bool `json:"include_running_balance,omitempty"`
}

func (o SyncOptions) body() map[string]any {
	body := map[string]any{}
	if o.Count > 0 {
		body["count"] = o.Count
	}
	if o.Cursor != "" {
		body["cursor"] = o.Cursor
	}
	if o.Options.IncludeRunningBalance {
		body["options"] = o.Options
	}
	return body
}

// Sync runs a transaction sync against an account.
// POST /v2/accounts/{id}/transactions/sync.
func (s *TransactionsService) Sync(ctx context.Context, accountID string, opts SyncOptions) (*tesote.TransactionSyncResponse, error) {
	out := &tesote.TransactionSyncResponse{}
	_, err := s.client.Do(ctx, "POST", pathPrefix+"/accounts/"+accountID+"/transactions/sync", tesote.RequestOptions{
		Body:           opts.body(),
		IdempotencyKey: opts.IdempotencyKey,
		Out:            out,
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

// SyncLegacy hits the non-nested legacy sync path.
// POST /v2/transactions/sync.
func (s *TransactionsService) SyncLegacy(ctx context.Context, opts SyncOptions) (*tesote.TransactionSyncResponse, error) {
	out := &tesote.TransactionSyncResponse{}
	_, err := s.client.Do(ctx, "POST", pathPrefix+"/transactions/sync", tesote.RequestOptions{
		Body:           opts.body(),
		IdempotencyKey: opts.IdempotencyKey,
		Out:            out,
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

// BulkOptions tunes POST /v2/transactions/bulk.
type BulkOptions struct {
	AccountIDs     []string
	Page           int
	PerPage        int
	Limit          int
	Offset         int
	IdempotencyKey string
}

// Bulk fetches transactions for multiple accounts in one call.
// POST /v2/transactions/bulk.
func (s *TransactionsService) Bulk(ctx context.Context, opts BulkOptions) (*tesote.BulkTransactionsResponse, error) {
	if len(opts.AccountIDs) == 0 {
		return nil, fmt.Errorf("tesote/v2: BulkOptions.AccountIDs must not be empty")
	}
	body := map[string]any{
		"account_ids": opts.AccountIDs,
	}
	if opts.Page > 0 {
		body["page"] = opts.Page
	}
	if opts.PerPage > 0 {
		body["per_page"] = opts.PerPage
	}
	if opts.Limit > 0 {
		body["limit"] = opts.Limit
	}
	if opts.Offset > 0 {
		body["offset"] = opts.Offset
	}
	out := &tesote.BulkTransactionsResponse{}
	_, err := s.client.Do(ctx, "POST", pathPrefix+"/transactions/bulk", tesote.RequestOptions{
		Body:           body,
		IdempotencyKey: opts.IdempotencyKey,
		Out:            out,
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

// SearchOptions tunes GET /v2/transactions/search.
type SearchOptions struct {
	Q         string
	AccountID string
	Limit     int
	Offset    int
	Filter    TransactionsFilter
}

// Search runs a text search across transactions. GET /v2/transactions/search.
func (s *TransactionsService) Search(ctx context.Context, opts SearchOptions) (*tesote.SearchTransactionsResponse, error) {
	if opts.Q == "" {
		return nil, fmt.Errorf("tesote/v2: SearchOptions.Q is required")
	}
	q := opts.Filter.query()
	q["q"] = opts.Q
	if opts.AccountID != "" {
		q["account_id"] = opts.AccountID
	}
	if opts.Limit > 0 {
		q["limit"] = strconv.Itoa(opts.Limit)
	}
	if opts.Offset > 0 {
		q["offset"] = strconv.Itoa(opts.Offset)
	}
	out := &tesote.SearchTransactionsResponse{}
	_, err := s.client.Do(ctx, "GET", pathPrefix+"/transactions/search", tesote.RequestOptions{
		Query: q,
		Out:   out,
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

File: packages/ts/src/v2/transactions.ts
import type { BulkResponse } from '../models/bulk.js';
import type { SearchResult } from '../models/search.js';
import type { SyncResult } from '../models/sync_transaction.js';
import type {
  CursorPagination,
  Transaction,
  TransactionListResponse,
} from '../models/transaction.js';
import type { Transport } from '../transport.js';

export interface V2TransactionListParams {
  start_date?: string;
  end_date?: string;
  scope?: string;
  page?: number;
  per_page?: number;
  transactions_after_id?: string;
  transactions_before_id?: string;
  transaction_date_after?: string;
  transaction_date_before?: string;
  created_after?: string;
  updated_after?: string;
  amount_min?: number;
  amount_max?: number;
  amount?: number;
  status?: string;
  category_id?: string;
  counterparty_id?: string;
  q?: string;
  type?: string;
  reference_code?: string;
}

export interface V2TransactionExportParams extends V2TransactionListParams {
  format?: 'csv' | 'json';
}

export interface ExportResponse {
  /** CSV body or pretty-printed JSON body, depending on `format`. */
  body: string;
  contentType: string | null;
  /** Suggested filename parsed from Content-Disposition (best-effort). */
  filename: string | null;
}

export interface SyncRequestOptions {
  count?: number;
  cursor?: string | 'now' | null;
  options?: {
    include_running_balance?: boolean;
  };
}

export interface BulkRequest {
  account_ids: string[];
  page?: number;
  per_page?: number;
  limit?: number;
  offset?: number;
}

export interface SearchParams extends V2TransactionListParams {
  q: string;
  account_id?: string;
  limit?: number;
  offset?: number;
}

const DISPOSITION_FILENAME = /filename\*?=(?:UTF-8'')?"?([^";]+)"?/i;

function parseFilename(disposition: string | null): string | null {
  if (disposition === null) return null;
  const match = DISPOSITION_FILENAME.exec(disposition);
  if (match === null) return null;
  const raw = match[1];
  if (raw === undefined) return null;
  try {
    return decodeURIComponent(raw);
  } catch {
    return raw;
  }
}

export class V2TransactionsClient {
  constructor(private readonly transport: Transport) {}

  /** GET /v2/accounts/{id}/transactions */
  async listForAccount(
    accountId: string,
    params: V2TransactionListParams = {},
  ): Promise<TransactionListResponse> {
    const res = await this.transport.request<TransactionListResponse>({
      method: 'GET',
      path: `/v2/accounts/${encodeURIComponent(accountId)}/transactions`,
      query: { ...params },
      cache: { ttl: 60 },
    });
    return res.data;
  }

  /** Cursor-following async iterator over /v2/accounts/{id}/transactions. */
  async *listAllForAccount(
    accountId: string,
    params: V2TransactionListParams = {},
  ): AsyncGenerator<Transaction, void, void> {
    let after = params.transactions_after_id;
    while (true) {
      const page = await this.listForAccount(accountId, {
        ...params,
        ...(after !== undefined ? { transactions_after_id: after } : {}),
      });
      for (const tx of page.transactions) yield tx;
      const pg: CursorPagination = page.pagination;
      if (!pg.has_more || pg.after_id === null) return;
      after = pg.after_id;
    }
  }

  /** GET /v2/transactions/{id} */
  async get(id: string): Promise<Transaction> {
    const res = await this.transport.request<Transaction>({
      method: 'GET',
      path: `/v2/transactions/${encodeURIComponent(id)}`,
      cache: { ttl: 300 },
    });
    return res.data;
  }

  /**
   * GET /v2/accounts/{id}/transactions/export — CSV or JSON file download.
   * The body is returned as a string; up to 10,000 transactions per call.
   */
  async export(accountId: string, params: V2TransactionExportParams = {}): Promise<ExportResponse> {
    const res = await this.transport.request<string>({
      method: 'GET',
      path: `/v2/accounts/${encodeURIComponent(accountId)}/transactions/export`,
      query: { ...params },
      headers: { Accept: '*/*' },
    });
    const contentType = res.headers.get('content-type');
    const body = typeof res.data === 'string' ? res.data : JSON.stringify(res.data);
    return {
      body,
      contentType,
      filename: parseFilename(res.headers.get('content-disposition')),
    };
  }

  /**
   * POST /v2/accounts/{id}/transactions/sync — Plaid-style flattened sync.
   */
  async sync(
    accountId: string,
    body: SyncRequestOptions = {},
    opts: { idempotencyKey?: string } = {},
  ): Promise<SyncResult> {
    const res = await this.transport.request<SyncResult>({
      method: 'POST',
      path: `/v2/accounts/${encodeURIComponent(accountId)}/transactions/sync`,
      body: { ...body },
      ...(opts.idempotencyKey !== undefined ? { idempotencyKey: opts.idempotencyKey } : {}),
    });
    return res.data;
  }

  /**
   * POST /v2/transactions/sync (legacy non-nested route). Identical request/response
   * to {@link sync} but takes account context in the body rather than the path.
   */
  async syncLegacy(
    body: SyncRequestOptions & { account_id?: string } = {},
    opts: { idempotencyKey?: string } = {},
  ): Promise<SyncResult> {
    const res = await this.transport.request<SyncResult>({
      method: 'POST',
      path: '/v2/transactions/sync',
      body: { ...body },
      ...(opts.idempotencyKey !== undefined ? { idempotencyKey: opts.idempotencyKey } : {}),
    });
    return res.data;
  }

  /** POST /v2/transactions/bulk */
  async bulk(body: BulkRequest, opts: { idempotencyKey?: string } = {}): Promise<BulkResponse> {
    const res = await this.transport.request<BulkResponse>({
      method: 'POST',
      path: '/v2/transactions/bulk',
      body,
      ...(opts.idempotencyKey !== undefined ? { idempotencyKey: opts.idempotencyKey } : {}),
    });
    return res.data;
  }

  /** GET /v2/transactions/search */
  async search(params: SearchParams): Promise<SearchResult> {
    const res = await this.transport.request<SearchResult>({
      method: 'GET',
      path: '/v2/transactions/search',
      query: { ...params },
    });
    return res.data;
  }
}
